File: ml-benchmark/synduce/treepaths/sum.f
Inductive BTree = empty Unit | node {Int, BTree, BTree};
Inductive Zipper = top Unit | left {Int, BTree, Zipper} | right {Int, BTree, Zipper};

sum = fix (
  \f: BTree -> Int. \t: BTree.
  match t with
    empty _ -> 0
  | node {a, l, r} -> + a (+ (f l) (f r))
  end
);



repr = fix (
  \f: Zipper -> Compress BTree. \z: Zipper.
  match z with
    top _ -> empty unit
  | left {w, tree, zz} ->
    /* (sum tree) is given in Synduce's template*/
    let tw = sum tree in
      node {w, tree, f zz}
  | right {w, tree, zz} ->
    let tw = sum tree in
      node {w, f zz, tree}
  end
);

main = \z: Zipper. sum (repr z);

File: ml-benchmark/synduce/treepaths/sum.ml
type btree = Empty of unit | Node of int * btree * btree
type zipper = Top of unit | Left of int * btree * zipper | Right of int * btree * zipper

let rec sum t =
  match t with
  | Empty _ -> 0
  | Node (a, l, r) -> a + sum l + sum r

let rec repr z =
  match z with
  | Top _ -> Empty ()
  | Left (w, tree, zz) ->
      let tw = sum tree in
      Node (w, tree, repr zz)
  | Right (w, tree, zz) ->
      let tw = sum tree in
      Node (w, repr zz, tree)

let program z = sum (repr z)